File: aemet/Aemet-Java-IvanRC/src/main/java/mapper/MedicionesMapper.java
package mapper;

import dto.ListaMedicionesDto;
import dto.MapaMedicionesDto;
import dto.MedicionDto;
import models.Medicion;
import utils.Utils;

import java.time.LocalDate;
import java.util.*;

public class MedicionesMapper {
    public static MedicionDto toDto(Medicion medicion){
        return new MedicionDto(
                medicion.getPoblacion(),
                medicion.getProvincia(),
                medicion.getTemperaturaMax().toString(),
                medicion.getHoraTempMax().toString(),
                medicion.getTemperaturaMin().toString(),
                medicion.getHoraTempMin().toString(),
                medicion.getPrecipitacion().toString(),
                medicion.getDia().toString()
        );
    }

    public static Medicion toMedicion(MedicionDto medicionDto){
        return new Medicion(
                medicionDto.getPoblacion(),
                medicionDto.getProvincia(),
                Double.parseDouble(medicionDto.getTemperaturaMax()),
                Utils.toLocalTime(medicionDto.getHoraTempMax()),
                Double.parseDouble(medicionDto.getTemperaturaMin()),
                Utils.toLocalTime(medicionDto.getHoraTempMin()),
                Double.parseDouble(medicionDto.getPrecipitacion()),
                Utils.toLocalDate(medicionDto.getDia())
        );
    }

    public static MapaMedicionesDto toMedicionesDto(Map<LocalDate, List<Medicion>> mapa){
        List<String> claves = mapa.keySet().stream().map(it -> it.toString()).toList();
        List<ListaMedicionesDto> valores = mapa.values().stream().map(it -> new ListaMedicionesDto(
                it.stream().map(medicion -> toDto(medicion)).toList()
        )).toList();
        Map<String, ListaMedicionesDto> mapas = new HashMap<>(Collections.emptyMap());
        for(int i = 0; i<claves.size(); i++){
            mapas.put(claves.get(i), valores.get(i));
        }
        return new MapaMedicionesDto(
                mapas
        );
    }

    public static Map<LocalDate, List<Medicion>> toMediciones(MapaMedicionesDto mapaDto){
        List<LocalDate> claves = mapaDto.getMapaMediciones().keySet().stream().map(it -> Utils.toLocalDate(it)).toList();
        List<List<Medicion>> valores = mapaDto.getMapaMediciones().values().stream().map(it ->
                it.getMediciones().stream().map(medicionDto -> toMedicion(medicionDto)).toList()
        ).toList();
        Map<LocalDate, List<Medicion>> mapas = new HashMap<>(Collections.emptyMap());
        for(int i = 0; i<claves.size(); i++){
            mapas.put(claves.get(i), valores.get(i));
        }
        return mapas;
    }
}

File: aemet/Aemet-kotlin-IvanRC/src/main/kotlin/mapper/MedicionesMapper.kt
package mapper

import dto.ListaMedicionesDto
import dto.MapaMedicionesDto
import dto.MedicionDto
import model.Medicion
import utils.toLocalDate
import utils.toLocalTime
import java.time.LocalDate

fun Medicion.toDto(): MedicionDto{
    return MedicionDto(
        poblacion = this.poblacion,
        provincia = this.provincia,
        temperaturaMax = this.temperaturaMax.toString(),
        horaTempMax = this.horaTempMax.toString(),
        temperaturaMin = this.temperaturaMin.toString(),
        horaTempMin = this.horaTempMin.toString(),
        precipitacion = this.precipitacion.toString(),
        dia = this.dia.toString()
    )
}

fun MedicionDto.toMedicion(): Medicion{
    return Medicion(
        poblacion = this.poblacion,
        provincia = this.provincia,
        temperaturaMax = this.temperaturaMax.toDouble(),
        horaTempMax = this.horaTempMax.toLocalTime(),
        temperaturaMin = this.temperaturaMin.toDouble(),
        horaTempMin = this.horaTempMin.toLocalTime(),
        precipitacion = this.precipitacion.toDouble(),
        dia = this.dia.toLocalDate()
    )
}

fun Map<LocalDate, List<Medicion>>.toMedicionesDto(): MapaMedicionesDto{
    return MapaMedicionesDto(
        this.mapKeys { it.key.toString() }.mapValues {medicion -> ListaMedicionesDto(medicion.value.map { it.toDto() }) }
    )
}

fun MapaMedicionesDto.toMediciones(): Map<LocalDate, List<Medicion>>{
    return this.medicionesDto
        .mapKeys { it.key.toLocalDate() }.mapValues { it.value.mediciones.map { it.toMedicion() } }
}